### DISCUSSION TITLE
Readwise imports with location

### DISCUSSION THREAD

**Discussion by CharlotteCMJ** [chhoumann/quickadd#385] [Q&A] [answered]
Hello, 

I am using QuickAdd and the EzImport script to import all my book highlights and notes from Readwise. I much prefer this solution to the official Readwise plugin as it gives me more freedom to use my Obsidian templates.

I was able to tweak the js file to take into account h2 tags on top of h1 tags. Furthermore, I mostly read on paper, and I like to add the page from which the highlight was taken. From the [Readwise API](https://readwise.io/api_deets), we can retrieve this information with the location_type and location keys. However, I am not familiar enough with Javascript to modify the code accordingly...

Here is the part of theEzImport script which, I believe, needs to be be modified :

```
function formatHighlights(highlights) {
    return highlights.map(hl => {
        if (hl.text == "No title") return;
        const {quote, note} = textFormatter(hl.text, hl.note);
        return `${quote}${note}`;
    }).join("\n\n");
}

function textFormatter(sourceText, sourceNote) {
    let quote = sourceText.split("\n").filter(line => line != "").map(line => {
        if (sourceNote.includes(".h1"))
            return `## ${line}`;
        else if (sourceNote.includes(".h2"))
            return `### ${line}`;
        else
            return `> ${line}`;
    }).join("\n");

    let note;

    if (sourceNote.includes(".h1") || sourceNote.includes(".h2") || sourceNote == "" || !sourceNote) {
        note = "";
    } else {
        note = "\n\n" + sourceNote;
    }

    return {quote, note};
}
```

Would anyone be able to help me out?

**Comment by chhoumann** [comment, marked as answer]
Hi @CharlotteCMJ

Glad to hear that you like EzImport.
You're correct that this is the right spot to modify. 

Whenever we've called the Readwise API to get highlights for a specific item, we get something like this back:
```json

{
  "count": 1163,
  "next": "https://readwise.io/api/v2/highlights?page=2",
  "previous": null,
  "results": [
    {
      "id": 59758950,
      "text": "The fundamental belief of metaphysicians is THE BELIEF IN ANTITHESES OF VALUES.",
      "note": "",
      "location": 9,
      "location_type": "order",
      "highlighted_at": null,
      "url": null,
      "color": "",
      "updated": "2020-10-01T12:58:44.716235Z",
      "book_id": 2608248,
      "tags": [
        {
           "id": 123456,
           "name": "philosophy"
        },
        ...
      ]
    },
    ...
  ]
}
```

You can e.g. edit `formatHighlights` to include details about the location of the highlight. E.g.
```javascript
function formatHighlights(highlights) {
    return highlights.map(hl => {
        if (hl.text == "No title") return;
        const {quote, note} = textFormatter(hl.text, hl.note);
        return `${quote}${note}\nPage: ${hl.location}\n`;
    }).join("\n\n");
}
```
Where `hl.location` refers to the location in the highlight (see `results` array items - the objects inside).
You embed that inside the template literals (strings) with the appropriate variables by writing `${hl.location}` as done above. Likewise for anything else in the object.

**Comment by CharlotteCMJ** [comment]
Hi Christian,

It works great and your explanation is very clear. Thank you so much for your help!

**Reply by chhoumann** [reply]
I'm glad to hear that. No problem!
